docs/writing-rules/glossary.md

```diff
@@ -0,0 +1,124 @@
+
+You can write checks for constant propagation through the use of `pattern-not` in conjunction with `pattern` or `metavariable-comparison`.
+
+It is enabled by default when writing rules and [can be disabled](/data-flow/constant-propagation/#disabling-constant-propagation).
+
+## Cross-file analysis
+
+Also known as **interfile analysis**. This analysis traces or tracks data and its transformations across files, such as when a variable is defined in one file but used in another.
+
+Cross file analysis can be paired with taint analysis to detect unsanitized variables flowing from a source to a sink.
+
+Also known as **interfile analysis**. Available in Semgrep Pro Engine.
+
+## Cross-function analysis
+
+This analysis traces or tracks data and its transformations across functions in a single file, such as when a globally-scoped variable is defined in one function but used in another.
+
+Also known as **intrafile** or **interprocedural** analysis. Available in Semgrep Pro Engine.
+
+## Error matrix
+
+An error matrix is a 2x2 table that visualizes the performance of a Semgrep rule in relation to the findings it does or doesn't detect. It has two axes:
+
+- Positive and negative
+- True or false
+
+These yield the following combinations:
+
+<dl>
+<dt>True positive</dt>
+<dd>The rule detected a piece of code it was intended to find.</dd>
+<dt>False positive</dt>
+<dd>The rule detected a piece of code it was not intended to find.</dd>
+<dt>True negative</dt>
+<dd>The rule correctly skipped over a piece of code it wasn't meant to find.</dd>
+<dt>False negative</dt>
+<dd>The rule failed to detect a piece of code it should have found.</dd>
+</dl>
+
+Not to be confused with **risk matrices**.
+
+## Finding
+
+A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. Findings can be security issues, bugs, or code that doesn't follow coding conventions.
+
+## Fully-qualified name
+
+A fully-qualified name refers to a name which uniquely identifies a class, method, type, or module. Languages such as C# and Ruby use `::` to distinguish between fully-qualified names and regular names.
+
+Not to be confused with **tokens**.
+
+## l-value (left-, or location-value)
+
+An expression that denotes an object in memory; a memory location, something that you can use in the left-hand side (LHS) of an assignment.
+
+Compare to [r-value](https://learn.microsoft.com/en-us/cpp/c-language/l-value-and-r-value-expressions?view=msvc-170).
+
+## Metavariable
+
+A metavariable is an abstraction that lets you match something even when you don't know exactly what it is you want to match. It is similar to capture groups in regular expressions. All metavariables begin with a `$` and can only contain uppercase characters, digits, and underscores.
+
+## Propagator
+
+A propagator is any code that alters a piece of data as the data moves across the program. This includes functions, reassignments, and so on.
```

Comment:
This definition might call into question why the user has to write their own propagators. I would say that in the context of a rule, propagators are pieces of code that you specify will always pass forward the tainted nature of data. This is particularly important in intraprocedural taint, where Semgrep cannot analyze other functions to figure out whether they will propagate taint or not. It also allows users to fine-tune Semgrep's analysis (because sometimes (often?) we get it wrong!)

Reply:
You're referring to having to spell out the functions / assignments in the `pattern-propagators` key, right?

I'm imagining that if it's intraprocedural taint, since it's within the function, there may be many assignments, so the engine would have to be told it's specifically `X = somePureFunction(Y)`?

I'll add your explanation in.

Reply:
![image](https://github.com/semgrep/semgrep-docs/assets/20766840/7576a855-ba13-4677-98a3-4a6fb2e1d9be)

Reply:
Definitely for intraprocedural taint you have to specify a lot of propagators yourself because we can't figure out if a function will act as a propagator.